### PR TITLE
8325858: Replace Unsafe usage in XEmbeddingContainer with FFM API

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to remove the use of `Unsafe` in the class `XEmbeddingContainer ` and replace it with the supported FFM API.

I tried to make this PR as small as possible while opening up for migration of other classes later on (such as `XEmbedServer` which can be modified similarly under a separate PR).

There are also three drive-by fixes in this PR:
 * Moved JavaDocs for `XAtom` to its proper location and fixed two typos in the text
 * Declared a field in `XEmbeddingContainer` as `private final`
 * In `XAtom`, use a utility method `assertAtomInitialized()` instead of the current duplicated code

Tested and passed tier1-5

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325858](https://bugs.openjdk.org/browse/JDK-8325858): Replace Unsafe usage in XEmbeddingContainer with FFM API (**Enhancement** - P5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17846/head:pull/17846` \
`$ git checkout pull/17846`

Update a local copy of the PR: \
`$ git checkout pull/17846` \
`$ git pull https://git.openjdk.org/jdk.git pull/17846/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17846`

View PR using the GUI difftool: \
`$ git pr show -t 17846`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17846.diff">https://git.openjdk.org/jdk/pull/17846.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17846#issuecomment-1943779516)

### CODE REVIEW

**Comment by mrserb** [comment]
Did you check how this change affects the cold startup? I think until the problem described in [1] is not fixed we cannot use the FFM. 
[1] https://bugs.openjdk.org/browse/JDK-8313344

**Comment by prrace** [comment]
In the interests of gauging what the actual impact might be (performance-wise), I tried looking for where we might create an XEmbeddingContainer() and I can find absolutely none. Making the sole constructor throw "new Error(..)" doesn't seem to cause any problems at all. This looks like dead code, unless someone else knows why we still have it.
The constructor XEmbedHelper() is called from some test code which is *in the source tree* not in a test.
However at least static fields and perhaps methods on that class are used.

But my conclusion from what I can see is that although the changes in XAtom seem like they could be re-used,
the best thing to do with XEmbeddingContainer might be to delete it.

Anyone have ANY idea where or how this might be used ? I'd like to be sure before I propose that.

FWIW the changes all look perfectly fine to me  .. although I am curious why a reachabilityFence was needed in one place. 

And the XAWT code has many uses of Unsafe.allocateMemory which could be candidates, but I assume that the timeline for removal of jdk.internal.misc.Unsafe (which is what this code uses) does not have to be the same as that for sun.misc.Unsafe memory allocation methods.

**Comment by mrserb** [comment]
>Anyone have ANY idea where or how this might be used ? I'd like to be sure before I propose that.

When the eawt was enhanced to support embedded frames all related non-public/internal code was not deleted - it was found that people used that and at that moment did not migrate to eawt(not sure that someone planned to do so since it worked).

**Comment by prrace** [comment]
> > Anyone have ANY idea where or how this might be used ? I'd like to be sure before I propose that.
> 
> When the eawt was enhanced to support embedded frames all related non-public/internal code was not deleted - it was found that people used that and at that moment did not migrate to eawt(not sure that someone planned to do so since it worked).

Embedded frames are for applets (dead) and JAWT and SWT. I don't see how any of those could use this.

Are you referring to people migrating AWAY from that Xembed stuff *TO* com.apple.eawt ?
I believe that the *very* early days of the macOS port had an X11 option but surely that's all dead and buried (as is com.apple.eawt) 
So I can't see how this code is needed or relevant today.

**Comment by mrserb** [comment]
> Embedded frames are for applets (dead) and JAWT and SWT. I don't see how any of those could use this.

yeah, I meant jawt, not eawt. SWT used some of the code directly from the native code, so they bypassed private access and modules. Might not be relevant anymore.

**Comment by prrace** [comment]
> > Embedded frames are for applets (dead) and JAWT and SWT. I don't see how any of those could use this.
> 
> yeah, I meant jawt, not eawt. SWT used some of the code directly from the native code, so they bypassed private access and modules. Might not be relevant anymore.

Ok, so sorry @minborg I think this is mostly pointless.
The changes in XAtom might be useful in migrating some of the other X code but for XEmbeddingContainer, the right answer seems to be to delete it. Not sure if you feel like that is your problem to tackle.
But probably the XAtom changes should be deferred to something that provably uses it in real apps which would reduce this PR to "delete obsolete code".

**Comment by prrace** [comment]
PS .. or it could just be all the drive-by clean up stuff which looked like it was aching to be done :-)

File: src/java.desktop/unix/classes/sun/awt/X11/XEmbeddingContainer.java [hunk 1]
```diff
@@ -130,19 +132,23 @@ public void dispatchEvent(XEvent xev) {
...
+                XKeyEvent ke = new XKeyEvent(data.address());
+                ke.set_window(child);
+                XToolkit.awtLock();
+                try {
+                    XlibWrapper.XSendEvent(XToolkit.getDisplay(), child, false, XConstants.NoEventMask, data.address());
+                } finally {
+                    XToolkit.awtUnlock();
+                }
+            } finally {
+                Reference.reachabilityFence(data);
```
Comment (bchristi-git) [inline src/java.desktop/unix/classes/sun/awt/X11/XEmbeddingContainer.java:150]: Can you explain the need for this `reachabilityFence()`? I don't understand how `data` could become unreachable before the `Arena` is closed. Also, it's an in-scope local variable. Thanks!

**Comment by minborg** [comment]
I ran an analysis tool that looked at about one million libraries (JARs) and there was just on occurrence of any of the three classes in user code: `de.mirkosertic.bytecoder:java.desktop:jar:2023-05-19`

File: src/java.desktop/unix/classes/sun/awt/X11/XEmbeddingContainer.java (hunk 1 above)
Comment (minborg) [reply src/java.desktop/unix/classes/sun/awt/X11/XEmbeddingContainer.java:150]: You are right. As we have the arena,  we do not need to invoke the reachabilityFence.

Comment (prrace) [reply src/java.desktop/unix/classes/sun/awt/X11/XEmbeddingContainer.java:150]: Yes, I had asked the same question !

**Pull request closed without merging**